File: pages/login_page.py
import time
from init_helpers import *

LOGIN_BUTTON_ELEMENT = ("role", "button", "Log in")
YOU_ARE_NOT_LOGGED_IN_LABEL_ELEMENT = ("text", "You are not logged in.")
ACCEPT_COOKIES_ELEMENT = ("role", "button", "I'm OK with analytics cookies")
DO_NOT_USE_ANALYTICS_COOKIES_ELEMENT = ("role", "button", "Do not use analytics cookies")
PAGE_LOADING_ELEMENT = ("role", "status")
TOTAL_VACCINATIONS_TEXT_ELEMENT = ("role", "heading", "Total vaccinations")
LOGOUT_NAV_ELEMENT = ("role", "link", "Log out")

def ensure_you_are_not_logged_in_label_exists():
    wait_for_element_to_disappear(PAGE_LOADING_ELEMENT)
    if not check_element_exists(YOU_ARE_NOT_LOGGED_IN_LABEL_ELEMENT):
        wait_for_element_to_disappear(PAGE_LOADING_ELEMENT)
        wait_for_element_to_appear(YOU_ARE_NOT_LOGGED_IN_LABEL_ELEMENT)

def navigate_to_ravs_login_page(url):
    navigate_to_url(url)
    if check_element_exists(ACCEPT_COOKIES_ELEMENT, False) is True:
        find_element_and_perform_action(ACCEPT_COOKIES_ELEMENT, "click")

def check_login_button_exists():
    return check_element_exists(LOGIN_BUTTON_ELEMENT, True)

def click_login_button():
    find_element_and_perform_action(LOGIN_BUTTON_ELEMENT, "click")
    wait_for_element_to_disappear(PAGE_LOADING_ELEMENT)

File: ts-tests/e2e/pages/login_page.ts
import { playwrightHelperInstance } from '../init_helpers';

const LOGIN_BUTTON_ELEMENT = { type: 'role', value: 'button', name: 'Log in' };
const YOU_ARE_NOT_LOGGED_IN_LABEL_ELEMENT = { type: 'text', value: 'You are not logged in.' };
const ACCEPT_COOKIES_ELEMENT = { type: 'role', value: 'button', name: "I'm OK with analytics cookies" };
const PAGE_LOADING_ELEMENT = { type: 'role', value: 'status' };
const LOGOUT_NAV_ELEMENT = { type: 'role', value: 'link', name: 'Log out' };

export async function ensureYouAreNotLoggedInLabelExists(): Promise<void> {
    await playwrightHelperInstance!.checkElementExists(YOU_ARE_NOT_LOGGED_IN_LABEL_ELEMENT);
}

export async function navigateToRavsLoginPage(url: string): Promise<void> {
  if (!playwrightHelperInstance) {
    console.error('playwrightHelperInstance is not initialized');
    throw new Error('playwrightHelperInstance is not initialized');
  }

  try {
    await playwrightHelperInstance!.navigateToUrl(url);
    await ensureYouAreNotLoggedInLabelExists();

      try {
      // Wait for the accept cookies element, and handle if it does not appear
      const acceptCookiesExists = await playwrightHelperInstance!.checkElementExists(ACCEPT_COOKIES_ELEMENT);

      if (acceptCookiesExists) {
        // Element found, perform the action
        await playwrightHelperInstance!.findElementAndPerformAction(ACCEPT_COOKIES_ELEMENT, 'click');
        console.log('Accepted cookies successfully.');
      } else {
        console.log('No accept cookies prompt found. Continuing with the test.');
      }
    } catch (error) {
      // Log error but don't interrupt the test if any error happens during checking element
      console.warn('No accept cookies element found or timed out. Continuing with the test...');
    }
  } catch (error) {
    // Log error but don't throw
    console.warn('Error while navigating or checking accept cookies element. Continuing with the test...');
  }
}

export async function checkLoginButtonExists(): Promise<boolean> {
    await ensureYouAreNotLoggedInLabelExists();
    return await playwrightHelperInstance!.checkElementExists(LOGIN_BUTTON_ELEMENT);
}

export async function clickLoginButton(): Promise<void> {
    await ensureYouAreNotLoggedInLabelExists();
    await playwrightHelperInstance!.findElementAndPerformAction(LOGIN_BUTTON_ELEMENT, 'click');
}
